provider/bidengine/pricing.go

```diff
@@ -60,9 +172,175 @@ func calculatePriceRange(gspec *dtypes.GroupSpec) (sdk.Coin, sdk.Coin) {
+type shellScriptPricing struct {
+	path         string
+	processLimit chan int
+	runtimeLimit uint
+}
+
+var errPathEmpty = errors.New("script path cannot be the empty string")
+var errProcessLimitZero = errors.New("process limit must be greater than zero")
+var errProcessRuntimeLimitZero = errors.New("process runtime limit must be greater than zero")
+
+func MakeShellScriptPricing(path string, processLimit uint, runtimeLimit uint) (BidPricingStrategy, error) {
+	if len(path) == 0 {
+		return nil, errPathEmpty
+	}
+	if processLimit == 0 {
+		return nil, errProcessLimitZero
+	}
+	if runtimeLimit == 0 {
+		return nil, errProcessRuntimeLimitZero
+	}
+
+	result := shellScriptPricing{
+		path:         path,
+		processLimit: make(chan int, processLimit),
+		runtimeLimit: runtimeLimit,
+	}
+
+	// Use the channel as a semaphore to limit the number of processes created for computing bid processes
+	// Most platforms put a limit on the number of processes a user can open. Even if the limit is high
+	// it isn't a good idea to open thuosands of processes.
+	for i := uint(0); i != processLimit; i++ {
+		result.processLimit <- 0
+	}
+
+	return result, nil
+}
+
+type dataForScriptElement struct {
+	Memory  uint64 `json:"memory"`
+	CPU     uint64 `json:"cpu"`
+	Storage uint64 `json:"storage"`
+	Count   uint32 `json:"count"`
+}
+
+var errCalculatePriceCancelled = errors.New("calculate price cancelled")
+var errCalculatePriceTimeout = errors.New("calculate price timed out")
+
+func (ssp shellScriptPricing) calculatePrice(gspec *dtypes.GroupSpec, lc lifecycle.Lifecycle) (sdk.Coin, error) {
+	buf := &bytes.Buffer{}
+
+	dataForScript := make([]dataForScriptElement, len(gspec.Resources))
+
+	// iterate over everything & sum it up
+	for i, group := range gspec.Resources {
+		groupCount := group.Count
+		cpuQuantity := group.Resources.CPU.Units.Val.Uint64()
+		memoryQuantity := group.Resources.Memory.Quantity.Value()
+		storageQuantity := group.Resources.Storage.Quantity.Val.Uint64()
+
+		dataForScript[i] = dataForScriptElement{
+			CPU:     cpuQuantity,
+			Memory:  memoryQuantity,
+			Storage: storageQuantity,
+			Count:   groupCount,
+		}
+	}
+
+	encoder := json.NewEncoder(buf)
+	err := encoder.Encode(dataForScript)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	cmd := exec.Command(ssp.path) //nolint:gosec
+	cmd.Stdin = buf
+	outputBuf := &bytes.Buffer{}
+	cmd.Stdout = outputBuf
+
+	cmdDone := make(chan error, 1)
+	// Take 1 from the channel
+	<-ssp.processLimit
+	defer func() {
+		// Always return it when this function is complete
+		ssp.processLimit <- 0
+	}()
+
+	err = cmd.Start()
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	go func() {
+		cmdDone <- cmd.Wait()
+	}()
+
+	halt := false
+	timeout := false
+	select {
+	case <-lc.ShuttingDown():
+		halt = true
+	case <-time.After(time.Duration(ssp.runtimeLimit) * time.Millisecond):
+		timeout = true
+	case err := <-cmdDone:
+		if err != nil {
+			return sdk.Coin{}, err
+		}
+
+	}
+
+	if halt || timeout {
```

Comment:
use [`CommandContext()`](https://golang.org/pkg/os/exec/#CommandContext) so you don't have to halt it yourself.

Reply:
> Looks great. Some comments inline.
> 
> I still think that if `shouldBid()` is false that we should be doing `break loop` instead of `break`.

If we do `break loop` when `shouldBid()` is false then that means no reservation is made and the goroutine associated with the `order` object terminates immediately after the first query comes back.

This is straightforward to implement. But what I don't understand is what the purpose of launching the goroutine in the first place? It's going to terminate basically immediately, the only side effect being the query that loads the group.

Reply:
It handles an order.  Goroutines are very cheap.